app/controllers/species_lists/projects_controller.rb

```diff
@@ -0,0 +1,191 @@
+# frozen_string_literal: true
+
+module SpeciesLists
+  class ProjectsController < ApplicationController
+    before_action :login_required
+    before_action :disable_link_prefetching
+
+    # ----------------------------
+    #  :section: Manage Projects
+    # ----------------------------
+
+    # def manage_projects
+    def edit
+      return unless (@list = find_species_list!)
+
+      if check_permission!(@list)
+        @projects = projects_to_manage
+        @object_states = manage_object_states
+        @project_states = manage_project_states
+      else
+        redirect_to(species_list_path(@list.id))
+      end
+    end
+
+    def update
+      return unless (@list = find_species_list!)
+
+      if check_permission!(@list)
+        @projects = projects_to_manage
+        @object_states = manage_object_states
+        @project_states = manage_project_states
+        commit_and_redirect
+      else
+        redirect_to(species_list_path(@list.id))
+      end
+    end
+
+    def projects_to_manage
+      projects = @user.projects_member
+      if @list.user == @user
+        projects += @list.projects
+        projects.uniq!
+      end
+      projects
+    end
+
+    def manage_object_states
+      {
+        list: params[:objects_list].present?,
+        obs: params[:objects_obs].present?,
+        img: params[:objects_img].present?
+      }
+    end
+
+    def manage_project_states
+      states = {}
+      @projects.each do |proj|
+        states[proj.id] = params["projects_#{proj.id}"].present?
+      end
+      states
+    end
+
+    def commit_and_redirect # rubocop:disable Metrics/AbcSize
+      case params[:commit]
+      when :ATTACH.l
+        if attach_objects_to_projects
+          redirect_to(species_list_path(@list.id)) and return
+        end
+
+        flash_warning(:runtime_no_changes.t)
+      when :REMOVE.l
+        if remove_objects_from_projects
+          redirect_to(species_list_path(@list.id)) and return
+        end
+
+        flash_warning(:runtime_no_changes.t)
+
+      else
+        flash_error("Invalid submit button: #{params[:commit].inspect}")
+      end
+      render(:edit)
+    end
+
+    def attach_objects_to_projects
```

Comment:
It would be nice to get rid the `@any_changes` flag. 
I haven't come up with anything yet. Maybe something for the future.

Reply:
Maybe have each of the sub-methods return a truthy value instead of setting a global flag.
And in `attach_objects_to_projects` have a local, memoized flag? 
There's still a flag, but at least it's local.
Still might be something for the future.

Reply:
If this logic moves to trailblazer, which i hope we can start doing soon, ivars like will be all turned into context args, to be passed between service objects, so i say let's leave it for now.

Reply:
Agree. Even if we don't go to Trailblazer soon, the refactor would (at least for me) require enough thought so that it best be done elsewhere.